lib/hexdocs/search/search.ex

```diff
@@ -0,0 +1,94 @@
+defmodule Hexdocs.Search do
+  require Logger
+
+  @type package :: String.t()
+  @type version :: Version.t()
+  @type proglang :: String.t()
+  @type search_items :: [map]
+
+  @callback index(package, version, proglang, search_items) :: :ok
+  @callback delete(package, version) :: :ok
+
+  defp impl, do: Application.fetch_env!(:hexdocs, :search_impl)
+
+  @spec index(package, version, proglang, search_items) :: :ok
+  def index(package, version, proglang, search_items) do
+    impl().index(package, version, proglang, search_items)
+  end
+
+  @spec delete(package, version) :: :ok
+  def delete(package, version) do
+    impl().delete(package, version)
+  end
+
+  @spec find_search_items(package, version, [{Path.t(), content :: iodata}]) ::
+          {proglang, search_items} | nil
+  def find_search_items(package, version, files) do
+    search_data_js =
+      Enum.find_value(files, fn {path, content} ->
+        case Path.basename(path) do
+          "search_data-" <> _digest -> content
```

Comment:
They can provide long strings like https://github.com/cloudpods-dev/docker-engine-api-elixir/blob/813cc557da483f623a8f484db04efc7e58db0376/lib/docker_engine_api/api/container.ex#L67, but Typesense seems to handle it fine. We can check for content size, maybe.

Reply:
I've added a test that checks that invalid fields in search items (like `type` being a map instead of a string, or `doc` being a list) are rejected: https://github.com/hexpm/hexdocs/pull/44/commits/8d58e4fbce076743d7b00e27ff595934a5d4360c